File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotEdition/tasks/PublishTask.java
package seigneur.gauvain.mycourt.ui.shotEdition.tasks;

import android.content.Context;
import android.net.Uri;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.HttpException;
import retrofit2.Response;
import seigneur.gauvain.mycourt.data.model.Shot;
import seigneur.gauvain.mycourt.data.repository.ShotDraftRepository;
import seigneur.gauvain.mycourt.data.repository.ShotRepository;
import seigneur.gauvain.mycourt.utils.ConnectivityReceiver;
import seigneur.gauvain.mycourt.utils.Constants;
import seigneur.gauvain.mycourt.utils.HttpUtils;
import seigneur.gauvain.mycourt.utils.image.ImageUtils;
import seigneur.gauvain.mycourt.utils.rx.NetworkErrorHandler;
import timber.log.Timber;

public class PublishTask {

    private CompositeDisposable mCompositeDisposable;

    private ShotRepository mShotRepository;

    private NetworkErrorHandler mNetworkErrorHandler;

    private ConnectivityReceiver mConnectivityReceiver;

    private PublishTaskCallBack mPublishTaskCallBack;

    private ShotDraftRepository mShotDraftRepository;

    @Inject
    public PublishTask(CompositeDisposable compositeDisposable,
                       ShotRepository shotRepository,
                       ShotDraftRepository shotDraftRepository,
                       NetworkErrorHandler networkErrorHandler,
                       ConnectivityReceiver connectivityReceiver,
                       PublishTaskCallBack publishTaskCallBack) {
        this.mCompositeDisposable=compositeDisposable;
        this.mShotRepository=shotRepository;
        this.mShotDraftRepository=shotDraftRepository;
        this.mNetworkErrorHandler=networkErrorHandler;
        this.mConnectivityReceiver=connectivityReceiver;
        this.mPublishTaskCallBack=publishTaskCallBack;
    }

    /*
    *************************************************************************
    * NETWORK OPERATION - POST SHOT ON DRIBBBLE
    *************************************************************************/
    private void postShot(Context context,
                          Uri fileUri,
                          String imageFormat,
                          String titleString,
                          String descriptionString,
                          ArrayList<String> tagList) {
        MultipartBody.Part body = HttpUtils.createFilePart(context,fileUri,imageFormat,"image");
        //add to HashMap key and RequestBody
        HashMap<String, RequestBody> map = new HashMap<>();
        // executes the request
        mCompositeDisposable.add(
                mShotRepository.publishANewShot(
                        map,
                        body,
                        titleString,
                        descriptionString,
                        tagList)
                        .subscribe(
                                this::onPostSucceed,
                                this::onPostFailed
                        )
        );
    }



    /**
     * Post operation has succeed - check the response from server
     * response must be 202 to be publish on Dribbble
     * @param response - body response from Dribbble
     */
    private void onPostSucceed(Response<Shot> response) {
        switch(response.code()) {
            case Constants.ACCEPTED:
                Timber.d("post succeed");
                //todo - do something in UI
                onPublishOrUpdateSucceed();
                break;
            default:
                Timber.d("post not succeed: "+response.code());
        }
    }

    /**
     * An error occurred while trying to make post request
     * @param t - throwable
     */
    private void onPostFailed(Throwable t) {
        Timber.e(t);
        handleNetworkOperationError(t,-1);
    }

    /*
    *************************************************************************
    * NETWORK OPERATION - UPDATE SHOT ON DRIBBBLE
    *************************************************************************/
    private void updateShot(
            String shotId,
            String title, String desc, ArrayList<String> tags, boolean profile) {
        mCompositeDisposable.add(
                mShotRepository.updateShot(
                        shotId, //get it from viewmodel
                        title,
                        desc,
                        tags,
                        profile)
                        .subscribe(
                                this::onUpdateShotSuccess,
                                this::onUpdateShotError
                        )
        );
    }

    /**
     * Shot update succeed
     * @param shot - shot updated
     */
    private void onUpdateShotSuccess(Shot shot) {
        //todo must finish with a code to send to Main Activity to delete the draft
        Timber.d("success: "+shot.getTitle());
        onPublishOrUpdateSucceed();
    }

    /**
     * Manage network error while trying to perform update
     * @param throwable - throwable
     */
    private void onUpdateShotError(Throwable throwable) {
        //todo - manage UI
        handleNetworkOperationError(throwable,-1);
    }

    /**
     * manage UI and DB items on Post/Updated Succeed
     */
    private void onPublishOrUpdateSucceed() {
        /*
        if (source==Constants.SOURCE_DRAFT) {
            //deleteDraft();  // reactivate
        }
        else {
            //todo listener
            //if (mEditShotView!=null)
            //    mEditShotView.stopActivity();
        }*/
    }


    /*
    *************************************************************************
    * DB OPERATION - DELETE DRAFT AFTER PUBLISH OR UPDATE
    *************************************************************************/
    private void deleteDraft() {
        /*mCompositeDisposable.add(
                mShotDraftRepository.deleteDraft(((ShotDraft) mObjectSource).getId())
                        .subscribe(
                                this::onDraftDeleted,
                                this::onDeleteDraftFailed
                        )
        );*/
    }

    /**
     * Draft has been deleted correctly
     */
    private void onDraftDeleted() {
        //mTempDataRepository.setDraftsChanged(true);//TODO LIVE DATA
        //TODO SINGLE LIVE EVENT
        /*if (mEditShotView!=null)
            mEditShotView.stopActivity();*/
    }

    /**
     * An error happened during delete process
     *
     * @param t - error description
     */
    private void onDeleteDraftFailed(Throwable t) {
        Timber.d(t);
    }

    /*
    *************************************************************************
    * MANAGE NETWORK EXCEPTION
    *************************************************************************/
    private void handleNetworkOperationError(final Throwable error, int eventID) {
        mNetworkErrorHandler.handleNetworkErrors(error,eventID, new NetworkErrorHandler.onRXErrorListener() {
            @Override
            public void onUnexpectedException(Throwable throwable) {
                Timber.d("unexpected error happened, don't know what to do...");
            }

            @Override
            public void onNetworkException(Throwable throwable) {
                Timber.d(throwable);
                if (mConnectivityReceiver.isOnline()) {
                    Timber.d("it seems that you have unexpected errors");
                } else {
                    Timber.d("Not connected to internet, so it is normal that you have an error");
                }

            }

            @Override
            public void onHttpException(Throwable throwable) {
                Timber.tag("HttpNetworks").d(throwable);
                if (((HttpException) throwable).code() == 403) {
                    //todo - access forbidden
                }
            }

        });
    }

    /**
     * CALLBACK FOR VIEWMODEL
     */
    public interface PublishTaskCallBack {

        void onTestGood();
    }


}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotEdition/tasks/PublishTask.kt
package seigneur.gauvain.mycourt.ui.shotEdition.tasks

import android.content.Context
import io.reactivex.disposables.CompositeDisposable
import seigneur.gauvain.mycourt.data.model.Attachment
import seigneur.gauvain.mycourt.data.model.Draft
import seigneur.gauvain.mycourt.data.repository.ShotDraftRepository
import seigneur.gauvain.mycourt.data.repository.ShotRepository
import seigneur.gauvain.mycourt.utils.ConnectivityReceiver
import seigneur.gauvain.mycourt.utils.Constants
import seigneur.gauvain.mycourt.utils.rx.NetworkErrorHandler
import timber.log.Timber


class PublishTask(
        private val mCompositeDisposable: CompositeDisposable,
        private val mShotRepository: ShotRepository,
        private val mShotDraftRepository: ShotDraftRepository,
        private val mNetworkErrorHandler: NetworkErrorHandler?,
        private val mConnectivityReceiver: ConnectivityReceiver,
        private val mPublishCallBack: PublishCallBack) {

    /*
    *************************************************************************
    * POST SHOT ON DRIBBBLE
    *************************************************************************/
    fun postShot(draft: Draft, context: Context) {
        mPublishCallBack.showWaiter(true)
        if (draft.hasAttachmentToPublish()) {
            //has attachment concat request
            mCompositeDisposable.add(
                    mShotRepository.postShotAndAttachment(draft, context)
                            .subscribe(
                                    { response ->
                                        when (response.code()) {
                                        Constants.ACCEPTED -> { onPublishSucceed(draft) }
                                        else ->
                                            onPublishFailed("Post failed: " + response.code() +": "+ response.message(), null)
                                        }},
                                    {t -> onPublishFailed(t.toString(), t)}
                            )
            )
        } else {
            mCompositeDisposable.add(
                    //doesn't have attachment
                    mShotRepository.publishANewShot(draft, context)
                            .subscribe(
                                    { response ->
                                        when (response.code()) {
                                            Constants.ACCEPTED -> { onPublishSucceed(draft) }
                                            else ->
                                                onPublishFailed("Post failed: " + response.code() +": "+ response.message(), null)
                                        }
                                    },
                                    {t -> onPublishFailed(t.toString(), t)}
                            )
            )
        }

    }


    /*
    *************************************************************************
    * UPDATE SHOT ON DRIBBBLE
    * 4 state possible :
    * A - only update Shot info
    * b - update shot info and post related attachment
    * c - update shot info and delete related attachment
    * d - update shot info, post and delete related attachment
    *************************************************************************/
    fun updateShot(
            draft: Draft,
            attachmentTodelete:List<Attachment>,
            context: Context,
            profile: Boolean) {
        mPublishCallBack.showWaiter(true)
        if (attachmentTodelete.isNotEmpty() && draft.hasAttachmentToPublish()) {
            //has new attachment to publish and old attachment to delete
            mCompositeDisposable.add(
                    mShotRepository.updateShotAndUploadAndDeleteAttachment(draft, profile, context, attachmentTodelete)
                            .subscribe(
                                    { _ -> Timber.d("update success updateShotAndUploadAndDeleteAttachment") },
                                    {  onPublishFailed("Update failed: $it",it)}
                            )
            )
        } else if (attachmentTodelete.isNotEmpty() && !draft.hasAttachmentToPublish()) {
            //has attachment to delete
            mCompositeDisposable.add(
                    mShotRepository.updateShotAndDeleteAttachment(draft, profile, attachmentTodelete)
                            .subscribe(
                                    { shot -> Timber.d("update success updateShotAndDeleteAttachment ")
                                        //onUpdateShotSuccess(shot, draft, context, attachmentTodelete)
                                    },
                                    { onPublishFailed("Update failed: $it", it)}
                            )
            )
        } else if (attachmentTodelete.isEmpty() && draft.hasAttachmentToPublish()) {
            //has new attachment to publish
            mCompositeDisposable.add(
                    mShotRepository.updateShotAndPostAttachment(draft, profile, context)
                            .subscribe(
                                    { shot -> Timber.d("update success updateShotAndPostAttachment")
                                        //onUpdateShotSuccess(shot, draft, context, attachmentTodelete)
                                    },
                                    {  onPublishFailed("Update failed: $it",it)}
                            )
            )
        } else {
            //no new attachment or delete attachment
            mCompositeDisposable.add(
                    mShotRepository.updateShot(draft, profile)
                            .subscribe(
                                    { shot -> Timber.d("updateShot success ")
                                        //onUpdateShotSuccess(shot, draft, context, attachmentTodelete)
                                    },
                                    {   onPublishFailed("Update failed: $it",it)
                                    }
                            )
            )
        }

    }

    /*
    *************************************************************************
    * DB OPERATION - DELETE DRAFT AFTER PUBLISH OR UPDATE
    *************************************************************************/
    private fun deleteDraftAfterPublish(draft:Draft) {
        mCompositeDisposable.add(mShotDraftRepository.deleteDraft(draft.draftID)
                .subscribe(
                        this::onDeleteSucceed,
                        this::onDeleteError
                )
        )

    }

    private fun onDeleteSucceed() {
        Timber.d("delete succeed")
    }

    private fun onDeleteError(throwable: Throwable) {
        Timber.e(throwable)
    }

    /*
    *************************************************************************
    * GLOBAL METHOD FOR EVERY REQUEST AND POST
    *************************************************************************/
    private fun onPublishSucceed(draft:Draft) {
        mPublishCallBack.showWaiter(false)
        deleteDraftAfterPublish(draft) //todo
        mPublishCallBack.onPublishSucceed()
    }

    private fun onPublishFailed(error:String, t:Throwable?) {
        t?.let{
            handleNetworkOperationError(t,100)
        }
        mPublishCallBack.showWaiter(false)
        mPublishCallBack.onPublishFailed(error)
    }

    /*
    *************************************************************************
    * MANAGE NETWORK EXCEPTION
    *
    * Note : we could use RxJavaPlugins.setErrorHandler but as we already
    * specified an error handler using subscribe, RxJavaPlugins.setErrorHandler will not be called.
    *************************************************************************/
    private fun handleNetworkOperationError(error: Throwable, eventID: Int) {
        Timber.d("error handled by rx")
        if (mNetworkErrorHandler == null) {
            Timber.tag("rxHandler").d("mNetworkErrorHandler is null")
        } else {
            mNetworkErrorHandler.handleNetworkErrors(error, eventID,
                    object : NetworkErrorHandler.onErrorListener {
                override fun onNetworkException(throwable: Throwable) {
                    if (mConnectivityReceiver.isOnline()) {
                        Timber.d("weird, you are connected, may be an interruption")
                    } else {
                        Timber.d("Not connected to internet")
                    }
                }
                override fun onHttpException(throwable: Throwable) {
                    Timber.d(throwable)
                }
            })
        }

    }

    /**
     * CALLBACK FOR VIEWMODEL
     */
    interface PublishCallBack {

        fun showWaiter(showIt:Boolean)

        fun onPublishSucceed()

        fun onPublishFailed(error:String)

    }

}
